Added actuator health endpoint and refactored unseal, seal-status, he…

singram opened the pull request:
…alth and initialize client support.

Resolves #24

mp911de commented on README.adoc:
```diff
@@ -163,6 +163,7 @@ spring.cloud.vault:
     scheme: http
     connection-timeout: 5000
     read-timeout: 15000
+    token: some-application-token
```
We did not mention the `token` here because authentication is a very individual configuration aspect.

singram replied:
Makes sense but without any authentication schema specified an application using this extension exits stating that an authentication token is mandatory.  I didn't see any mention of other authentication schemes in the docs (only in the code) and added as a immediate assist for new comers (like myself :-)) most likely to use the extension with Vaults primary authentication scheme.  Agreed, this is probably best broken out into a separate section on authentication.

mp911de replied:
Right, this document does not mention authentication at all. I created a new ticket to improve the quickstart guide. We have docs on authentication in [spring-cloud-vault-config.adoc](https://github.com/spring-cloud-incubator/spring-cloud-vault-config/blob/master/docs/src/main/asciidoc/spring-cloud-vault-config.adoc#authentication-methods)

singram replied:
Awesome!  I completely missed this documentation, thanks!  Might be worth noting that the 'authentication' key for authentication method is optional and seems to default to 'TOKEN'.

mp911de commented on README.adoc:
```diff
@@ -172,10 +173,17 @@ for SSL certificate validation
 Supported schemes are `http` and `https`.
 * `connection-timeout` sets the connection timeout in milliseconds
 * `read-timeout` sets the read timeout in milliseconds
+* `token` sets the main token to authenticate with vault.  This can also be set with the environment variable SPRING_CLOUD_VAULT_TOKEN
```
Mentioning the environment variable fits good to the token authentication section.

singram replied:
Agreed.  I'd missed the appropriate documentation spot to add this

mp911de replied:
I missed one point, sorry for that: `README.md` is generated from `docs/src/main/asciidoc/README.adoc` using `docs/src/main/ruby/generate_readme.sh`. If you want to contribute documentation changes, please do these in the documentation sources in `docs/src/main/asciidoc/`. I added a reference to the reference documentation in the meantime.

singram replied:
No problem at all, thanks for the heads up.  I should have something up in the next day or so.  Free time this week has been in short supply, my apologies for being unable to post a revision sooner.

mp911de commented on README.adoc:
```diff
@@ -190,11 +198,16 @@ setup are scripted, the scripts are located in `src/test/bash`.
 ----
+$ ./src/test/bash/install_vault.sh
+$ ./src/test/bash/create_certificates.sh
+$ ./src/test/bash/local_run_vault.sh
```
👍

mp911de commented on spring-cloud-vault-core/src/main/java/org/springframework/cloud/vault/VaultClient.java:
```diff
@@ -102,6 +124,94 @@ public VaultClientResponse write(URI uri, Object entity, VaultToken vaultToken)
+	 * @param requiredKeys The number of keys required to unseal
+	 * @return A {@link VaultInitializedResponse} containing the initialized state of vault.
+	 */
+	public VaultInitializedResponse initialize(int createKeys, int requiredKeys) {
```
These features rather belong into the `VaultTemplate`. Let's focus on #24 and keep the other changes out of the PR (see explanation in the summary).

mp911de commented on spring-cloud-vault-core/src/main/java/org/springframework/cloud/vault/VaultClient.java:
```diff
@@ -102,6 +124,94 @@ public VaultClientResponse write(URI uri, Object entity, VaultToken vaultToken)
+	 *
+	 * @return A {@link VaultHealthResponse} containing the current service status.
+	 */
+	public VaultHealthResponse health() {
```
I think the health check could be moved inside the `VaultHealthIndicator`

singram replied:
I'm not sure I agree.  As a user of the extension I would like to interact with Vault through a single client (my assumption is that this would be VaultClient).  In the absence of using Actuator or even outside of the HealthIndicator I may still find need to check the health of the Vault service.  I think this plays into a discussion at the end of the PR.

mp911de replied:
I should have been clearer, sorry. My intention is to obtain the health response using the `VaultClient` in `VaultHealthIndicator`. You're totally right that one should not use directly `RestTemplate`.

singram replied:
Ah cool.  It does that already (line 33), are there further changes you're requesting?

mp911de replied:
By removing `VaultProperties` and the other non-related changes from this class, you need to pass in the `URI` to specify the host/port/path. `VaultClient` is not configured to a specific host but just wraps the protocol.

singram replied:
VaultTemplate wraps the properties and client and per comments below.  However this is in the config package, would it be acceptable to bring this back into the 'core' package as the first order vault interaction object?  The config package then would be focused entirely on spring property/configuration management.

mp911de replied:
That's exactly part of our future refactoring. We will do so once your PR is merged.

mp911de commented on spring-cloud-vault-core/src/main/java/org/springframework/cloud/vault/VaultClient.java:
```diff
@@ -170,6 +291,10 @@ private HttpHeaders createHeaders(VaultToken vaultToken) {
 		return headers;
 	}
 
+	private HttpHeaders createHeaders() {
```
`VaultClient` is login-method agnostic. All state and login features should be kept inside `VaultTemplate`. We support various login methods that keep their state in a `VaultState` class. It's not intended to write the token back into `VaultProperties`.

singram replied:
Good point I wasn't considering other authentication mechanisms in this PR.

singram replied:
I'd also missed this (VaultTemplate) as my assumption (sorry, made a lot of those :)) was that everything needed for this change was in the core package.

mp911de commented on spring-cloud-vault-core/src/main/java/org/springframework/cloud/vault/VaultInitializeRequest.java:
```diff
@@ -0,0 +1,33 @@
+/*
```
Let's skip the unseal/initialization refactorings for this PR.

mp911de commented on spring-cloud-vault-core/src/main/java/org/springframework/cloud/vault/VaultInitializedResponse.java:
```diff
@@ -0,0 +1,34 @@
+/*
```
Let's skip the unseal/initialization refactorings for this PR.

mp911de commented on spring-cloud-vault-core/src/main/java/org/springframework/cloud/vault/VaultResponse.java:
```diff
@@ -16,11 +16,11 @@
 package org.springframework.cloud.vault;
 
-import java.util.Map;
+import lombok.Data;
```
Any reason for the reordering of imports?

singram replied:
Absolutely none.  I imported the eclipse style sheet specified and this happened.  Will revert.  Any configuration I could import to align with your standards?

mp911de replied:
Here's the link to [eclipse-code-formatter.xml](https://github.com/spring-cloud/spring-cloud-build/blob/master/spring-cloud-dependencies-parent/eclipse-code-formatter.xml).

singram replied:
Thanks but that's what I imported.  Import ordering seems to have a separate import/export from formatting (Java->Code Style->Organize Imports in Eclipse).  Will keep a closer eye on import changes moving forwards

mp911de commented on spring-cloud-vault-core/src/main/java/org/springframework/cloud/vault/VaultSealStatusResponse.java:
```diff
@@ -0,0 +1,33 @@
+/*
```
Let's skip the unseal/initialization refactorings for this PR.

mp911de commented on spring-cloud-vault-core/src/test/java/org/springframework/cloud/vault/CertAuthenticationMethodsIntegrationTests.java:
```diff
@@ -15,10 +15,14 @@
  */
```
Let's skip the unseal/initialization refactorings for this PR.

mp911de commented on spring-cloud-vault-core/src/test/java/org/springframework/cloud/vault/ClientHttpRequestFactoryFactoryIntegrationTests.java:
```diff
@@ -15,69 +15,53 @@
  */
```
Let's skip the unseal/initialization refactorings for this PR.

mp911de commented on spring-cloud-vault-core/src/test/java/org/springframework/cloud/vault/util/PrepareVault.java:
```diff
@@ -16,45 +16,42 @@
 
```
Let's skip the unseal/initialization refactorings for this PR.

mp911de commented:
Thanks @singram for the PR. This PR contains several neat changes which make all sense if considered alone. The code looks really good, but some parts don't fit the overall design. The idea behind `VaultClient` is to provide a protocol abstraction for Vault (request/response/error handling). Higher level functions are usually part of `VaultOperations` (`VaultTemplate` as implementation). `VaultClient` is also opinionated on a login method. The login happens in `VaultTemplate` as that's the component holding the state.

We're considering to extract the core Vault support into an own module so other Spring projects can benefit from the core, but this requires a refactoring on our side. We will add dedicated support for specific Vault features (Initialization, (Un)Sealing, Revocation, Token management, …) so your changes would be useful in that context. I would propose to keep the actuator support change in the PR and postpone the refactorings. 

Does this make sense?

singram commented:
@mp911de  Thanks for the great feedback it's appreciated!  As a new user a lot of the design was unclear.  My assumption was that VaultClient is the primary class that would interact with Vault API's, hence the migration of API interactions to that class along with configuration properties.  It's also in the Core package which may be used (my assumption completely!) outside of the Spring configuration property injection functionality currently implemented.  Admittedly in the changes I completely neglected to consider other authentication methods doh!

Happy to help in any refactoring or documentation.  As I said, I made a number of assumptions and missed the multiple authentication methods.

Perhaps an example for how a developer is supposed to interact with the client would help others seeking to use this project, I know there's a significant amount of interest in this project and Vault.

spencergibb commented on spring-cloud-vault-core/pom.xml:
```diff
@@ -22,6 +22,11 @@
 			<artifactId>spring-cloud-context</artifactId>
 		</dependency>
 
+        <dependency>
```
We use tabs for formatting here.

singram replied:
Thanks @spencergibb.  I'll adjust on the next update.  Is there configuration beyond the code formatter mentioned in the docs I can apply to better comply with your standards?

singram commented:
Thanks again for all the feedback.  I'll push up a more focused update in the next day or so addressing all the comments.

mp911de commented on spring-cloud-vault-core/src/main/java/org/springframework/cloud/vault/VaultClient.java:
```diff
@@ -102,6 +124,94 @@ public VaultClientResponse write(URI uri, Object entity, VaultToken vaultToken)
+	 *
+	 * @return
+	 */
+	public boolean isAvailable() {
```
The only subscriber for health is the `VaultHealthIndicator` so `isAvailable()` should move into the health indicator.

singram replied:
This seems the be a useful abstraction for the test setup as well.  I left it on the client to dry up some of the test code and promote the execution of client code in the tests.

mp911de commented on spring-cloud-vault-core/src/main/java/org/springframework/cloud/vault/VaultClient.java:
```diff
@@ -15,45 +15,67 @@
 
-	public VaultClient() {
-		this(new RestTemplate());
+	@Getter
```
`VaultProperties` and the changed constructors do not belong here. The only configuration for `VaultClient` is `RestTemplate`.

singram replied:
Agreed.  I'll look to the VaultTemplate to facilitate this functionality moving forwards.

mp911de commented:
I added some more comments as answer to your questions.

singram commented:
Cleaned up version pushed which hopefully takes into account all of the excellent feedback.
I moved the HealthIndicator into the 'config' package as this seemed closer to spring-boot usage, assuming the 'core' package will become more of a general java-vault client interface.

mp911de commented:
Awesome! Thanks a lot. That's merged and polished with `d1e710b`.
